### DISCUSSION TITLE
Weird noise on saving patch (owm external)

### DISCUSSION THREAD

**Discussion by charlesneimog** [pure-data/pure-data#2222] [Q&A] [answered]
@umlaeute, @porres, @Spacechild1

Sorry to ask about a personal project, but I am having a weird noise when I save the patch of using my external. I can't figure out what is the problem.

I am not working with `binbuf` or something like it. 

The noise when I save it's ok, my problem it is the persistent noise (end of the video).

https://github.com/pure-data/pure-data/assets/31707161/87d7042e-6baf-4b21-a59b-d9930b396de5

You have some idea what it can be?

Someone had some similar problem?

https://github.com/charlesneimog/pd-partialtrack

To build:
`git clone https://github.com/charlesneimog/pd-partialtrack  --recursive && cd pd-partialtrack`
`cmake . -B build && cmake --build build` 

Thanks for any help!!

**Comment by Spacechild1** [comment, marked as answer]
Saving a patch triggers a DSP graph update, so I would assume that the problems have to do with the "dsp" method and/or perform method.

---

I had a quick glance at the code of `s-peaks~` and `s-synth~` and I noticed some serious issues:

https://github.com/charlesneimog/pd-partialtrack/blob/d82a2f5518a04f8f308c70120ed35cd52ad538cf/Sources/s-peaks~.cpp#L167-L172

Never ever send a Pd message from within the perform routine! For every scheduler tick, Pd first does all the message handling and then all the DSP; these two systems must not interleave! If you want to output a message as a result of DSP computation, you must set a clock, typically with `clock_delay(clock, 0)`. Of course, the clock tick method will only be called *after* all the DSP has finished. If that message is supposed to interact with other DSP objects, you will get a delay of 1 block. There is no (safe) way around this!

https://github.com/charlesneimog/pd-partialtrack/blob/d82a2f5518a04f8f308c70120ed35cd52ad538cf/Sources/s-peaks~.cpp#L173-L178

This creates a new thread for every `BufferSize` samples 😱 Of course, you shouldn't be spawning threads in the perform routine in the first place. Multithreading is a complex topic. Honestly, I would advice you not to write multithreaded code unless you really know what you are doing.

https://github.com/charlesneimog/pd-partialtrack/blob/d82a2f5518a04f8f308c70120ed35cd52ad538cf/Sources/s-synth~.cpp#L114-L117

You seem to passing around raw C++ pointers via Pd messages as `t_gpointer *` arguments. That is dangerous as hell! If you want to share data between Pd objects, you should just bind them to symbols. Then individual objects can reference each other with `pd_findbyclass`.

https://github.com/charlesneimog/pd-partialtrack/blob/d82a2f5518a04f8f308c70120ed35cd52ad538cf/Sources/s-peaks~.cpp#L256
https://github.com/charlesneimog/pd-partialtrack/blob/d82a2f5518a04f8f308c70120ed35cd52ad538cf/Sources/s-synth~.cpp#L135-L138

The memory is never freed!

Hint: your Pd object does not have to be a POD, they can be full C++ classes (but no *virtual* member functions!) The "new" function would just allocate the object with `pd_new` and then call placement-new. The "free" function just manually calls the destructor. Then you can have C++ classes as members. For example, the buffer above could be a `std::vector<t_sample>`.

---

In general, it looks as if you tried to build a modular set of DSP objects that can interact with each other For this purpose you are currently abusing the message system. The proper solution would be to use symbols, similar to `[delwrite~]` / `[delread~]`, `[s~]` / `[r~]`, `[tabwrite~]` / `[tabread~]`, etc.

**Comment by charlesneimog** [comment]
Amazing,

Thank you very much for your time, a lot of great info for me!!!

**Comment by charlesneimog** [comment]
@Spacechild1, I have another question about raw data using `t_gpointer`. 

In this library seems ok (patching perspective) to refer to `peaks~` and `synth~` using names etc. But I am using the same process for `py4pd`, `py4pd` allows to use of any Python data type inside pure data. But it is not possible to make this work without using `g_pointer` (which was my unique solution). 

https://github.com/charlesneimog/py4pd

There is some way that allows me to use this kind of thing? Basically, I need to pass a pointer to a `PyObject` struct between objects.  And I'd like it to work as normal patching. Always using objects 'connections' by names is not the better way for `py4pd`.

What I understand is that, if the user does not use `unpack s s p` (I use messages like `PyObject <pyclassname (symbol)> (pointer)`) to get just the pointer, this will not generate major problems. I'm just fooling PureData.

Am I veryyy wrong?

Thank you again!

**Reply by Spacechild1** [reply]
> But it is not possible to make this work without using g_pointer (which was my unique solution).

Why not? You can really bind *anything* to a Pd symbol, you just need to wrap it in a fake class, so it can be found with `pd_findbyclass`.

> Basically, I need to pass a pointer to a PyObject struct between objects.

Instead of passing around the raw data as fake gpointers (*shudder*), just bind them to unique symbols and pass around the symbols. Note that the symbols only need to be unique *per class*. In fact, you could just use the memory address itself as a string. (This is actually how Pd objects are identified on the GUI side.)

> Am I veryyy wrong?

I am afraid the answer is likely yes :) As a rule of thumb: Pd externals should be *safe*, i.e. they should never be able to crash at runtime.

**Reply by Spacechild1** [reply]
In general, if you find yourself doing something really funky and you are not 100% sure it is the only solution, play it safe and ask on the Pd dev list beforehand. You will typically get practical advice in the next 24h hours :)

**Reply by charlesneimog** [reply]
Great, symbol with data address seemed more wrong for me hheheh... I was wrong then... 

Big thanks for the help again!  :)

**Reply by Spacechild1** [reply]
You're welcome! I really think of Pd symbols as safe named resource handles.

**Reply by charlesneimog** [reply]
I will change to symbols to represent the actual pointer as a string. It seems more simple to me... Thanks!

**Reply by Spacechild1** [reply]
> I will change to symbols to represent the actual pointer as a string.

I hope that you don't plan to just (de)serialize the memory address and cast it back to a pointer. That would be fundamentally unsafe because there is no guarantee that the pointer is still valid.

You absolutely need to wrap your resource in a proxy object so that clients can *safely* obtain it with `pd_findbyclass`. Needless to say, clients must not store a direct pointer to the underlying data; instead they should store the symbol and use `pd_findbyclass` everytime they want to access it. It is really the same as with graphical arrays!

Here's a quick example:

```
t_class *foo_class;

typedef struct _foo
{
    t_pd x_pd; // only a class pointer
    t_symbol *x_sym; // the symbol this object is bound to
    Foo *x_data; // the actual resource
} t_foo;

// create Foo instance and wrap inside t_foo object
t_foo *create_foo(void)
{
    char buf[64];
    t_foo*x = (t_foo*)getbytes(sizeof(t_foo));
    x->x_pd = foo_class;
    x->x_data = Foo_new(); // create the actual resource
    snprintf(buf, sizeof(buf), "%p", x); // just use the memory address
    x->x_sym = gensym(buf);
    pd_bind((t_pd *)x, x->x_sym);
}

// destroy the t_foo object and its containing Foo resource
void destroy_foo(t_foo *x)
{
    pd_unbind((t_pd *)x, x->x_sym);
    Foo_destroy(x->x_data); // destroy the actual resource
    freebytes(x);
}

// get Foo instance by name; returns 0 if it does not exist (anymore)
Foo *get_foo(t_symbol *s)
{
    t_foo *x = pd_findbyclass(s, foo_class);
    return x ? x->x_data : 0; 
}

my_proxy_class = class_new(gensym("my proxy class"), 0, 0, sizeof(t_node_proxy), CLASS_PD, A_NULL);
```

**Reply by charlesneimog** [reply]
Ahhh, now I see, I will try to be careful with pointers.

Anyway, thank for the help, it helps me a lot!
